File: collection_musique/vue.cpp
#include <QtGui>
#include <QtSql>

#include "albummodel.h"
#include "connexion.h"
#include "csvparser.h"
#include "defaulttableview.h"
#include "login.h"
#include "sqlparser.h"
#include "usersview.h"
#include "vue.h"

Vue::Vue(Controleur *cont, QWidget *parent) : QMainWindow(parent) {
    init();
}

Vue::~Vue() {
    delete _btnList;
    delete _btnNbAlbums;
    delete _btnInfos;
    delete _btnUsers;
    delete _btnDelete;
    delete _layBoutons;
    delete _tabView;
    delete _model;
    delete _centralLayout;
    delete _centralWidget;
}

void Vue::init() {
    //Connexion a la base
    openSqlConnection();
    
    #ifndef DEBUG
    Login l(this);
    if (l.exec() == 0)
        exit(-1);
    #endif
    
    _centralWidget = new QWidget();
    this->setCentralWidget(_centralWidget);
    this->setWindowTitle(tr("Collection de musique"));
    
    _centralLayout = new QVBoxLayout();
    
    //Creation des menus
    _mnuFile = new QMenu(tr("&File"), this);
    _mnuImport = new QMenu(tr("&import"), this);
    
    _actImportCSV = new QAction(tr("CSV File"), this);
    _actImportSQL = new QAction(tr("SQL File"), this);
    _actQuit = new QAction(tr("&Quit"), this);
    
    connect(_actImportCSV, SIGNAL(triggered()), this, SLOT(importCSV()));
    connect(_actImportSQL, SIGNAL(triggered()), this, SLOT(importSQL()));
    connect(_actQuit, SIGNAL(triggered()), qApp, SLOT(quit()));
    
    menuBar()->addMenu(_mnuFile);
    _mnuFile->addMenu(_mnuImport);
    _mnuFile->addAction(_actQuit);
    _mnuImport->addAction(_actImportCSV);
    _mnuImport->addAction(_actImportSQL);
    
    //Ajout des boutons
    _btnList = new QPushButton(tr("Liste"));
    _btnNbAlbums = new QPushButton(tr("Nb. Albums"));
    _btnInfos = new QPushButton(tr("Infos"));
    _btnUsers = new QPushButton(tr("Users"));
    _btnDelete = new QPushButton(tr("Delete"));
    connect(_btnList, SIGNAL(clicked()), this, SLOT(showList()));
    connect(_btnNbAlbums, SIGNAL(clicked()), this, SLOT(showNbAlbums()));
    connect(_btnInfos, SIGNAL(clicked()), this, SLOT(showInfos()));
    connect(_btnUsers, SIGNAL(clicked()), this, SLOT(showUsers()));
    connect(_btnDelete, SIGNAL(clicked()), this, SLOT(emptyTable()));
    
    _layBoutons = new QHBoxLayout();
    _layBoutons->addWidget(_btnList);
    _layBoutons->addWidget(_btnNbAlbums);
    _layBoutons->addWidget(_btnInfos);
    _layBoutons->addWidget(_btnUsers);
    _layBoutons->addWidget(_btnDelete);
    
    _centralLayout->addLayout(_layBoutons);
    
    //Modèle initial
    _model = new AlbumModel();
    //~ showList();
    
    //Setup de la table (fixe)
    _tabView = new DefaultTableView();
    _tabView->setModel(_model);
    _tabView->resizeColumnsToContents();
    _tabView->sortByColumn(1, Qt::AscendingOrder);
    _centralLayout->addWidget(_tabView);
    
    _centralWidget->setLayout(_centralLayout);
    
    this->setMinimumSize(800, 600);
    this->show();
}

void Vue::showList() {
    _query = "SELECT id, artiste, album, annee, genre, pistes, time(duree, 'unixepoch') FROM collection ORDER BY artiste";
    _model->setQuery(_query);
    _model->setHeaderData(0, Qt::Horizontal, tr("Id"));
    _model->setHeaderData(1, Qt::Horizontal, tr("Artiste"));
    _model->setHeaderData(2, Qt::Horizontal, tr("Album"));
    _model->setHeaderData(3, Qt::Horizontal, tr("Annee"));
    _model->setHeaderData(4, Qt::Horizontal, tr("Genre"));
    _model->setHeaderData(5, Qt::Horizontal, tr("Pistes"));
    _model->setHeaderData(6, Qt::Horizontal, tr("Duree"));
}

void Vue::showNbAlbums() {
    _query = "SELECT artiste, count(*), min(annee), sum(pistes), strftime('%w %H:%M:%S', sum(duree), 'unixepoch', '-4 days'), time(sum(duree) / sum(pistes), 'unixepoch') moy FROM collection GROUP BY artiste ORDER BY artiste";
    _model->setQuery(_query);
    _model->setHeaderData(0, Qt::Horizontal, tr("Artiste"));
    _model->setHeaderData(1, Qt::Horizontal, tr("Nb. Albums"));
    _model->setHeaderData(2, Qt::Horizontal, tr("1ere Annee"));
    _model->setHeaderData(3, Qt::Horizontal, tr("Pistes"));
    _model->setHeaderData(4, Qt::Horizontal, tr("Duree"));
    _model->setHeaderData(5, Qt::Horizontal, tr("Moy Chanson"));
}

void Vue::showInfos() {
    _query = "SELECT count(distinct artiste) 'Nb Artistes', count(album) 'Nb Albums', count(distinct genre) 'Nb Genres', sum(pistes) 'Nb Pistes', strftime('%j %H:%M:%S', sum(duree), 'unixepoch', '-1 days') 'Duree tot', time(sum(duree) / sum(pistes), 'unixepoch', '-4 days') 'Moy long chanson' FROM collection";
    _model->setQuery(_query);
    _model->setHeaderData(0, Qt::Horizontal, tr("Nb Artistes"));
    _model->setHeaderData(1, Qt::Horizontal, tr("Nb Albums"));
    _model->setHeaderData(2, Qt::Horizontal, tr("Nb Genres"));
    _model->setHeaderData(3, Qt::Horizontal, tr("Nb Pistes"));
    _model->setHeaderData(4, Qt::Horizontal, tr("Duree tot"));
    _model->setHeaderData(5, Qt::Horizontal, tr("Moy long chanson"));
}

void Vue::showUsers() {
    UsersView v;
    v.exec();
}

void Vue::emptyTable() {
    QSqlQuery q;
    q.exec("DELETE FROM collection");
    showList();
}

void Vue::importCSV() {
    CsvParser csv(this);
    csv.import();
    showList();
}

void Vue::importSQL() {
    SqlParser sql(this);
    sql.import();
    showList();
}

File: collection_musique/vue.py
#! /usr/bin/env python
#-*- coding: UTF-8 -*-

from PyQt4.QtCore import *
from PyQt4.QtGui import *
from PyQt4.QtSql import *
import sys

import ajoutalbum
import connexion
import customsqlmodel as csm

#Raccourcis
tr = QObject.tr

class Vue(QMainWindow):
    def __init__(self, cont, parent=None):
        super(Vue, self).__init__(parent)
        self.centralWidget = QWidget()
        self.setCentralWidget(self.centralWidget)
        self.setWindowTitle('Collection de musique')
        self.setMinimumSize(800, 600)
        
        self.centralLayout = QVBoxLayout()
        
        #Ajout des boutons
        self.btnAll = QPushButton(tr(self, "Tout"))
        self.btnNbAlbums = QPushButton(tr(self, "Nb. Albums"))
        self.btnAdd = QPushButton(tr(self, "Ajout"))
        self.btnQuit = QPushButton(tr(self, "Quitter"))
        self.connect(self.btnAll, SIGNAL('clicked()'), self.showAll)
        self.connect(self.btnNbAlbums, SIGNAL('clicked()'), self.showNbAlbums)
        self.connect(self.btnAdd, SIGNAL('clicked()'), self.ajout)
        self.connect(self.btnQuit, SIGNAL('clicked()'), self, SLOT('close()'))
        
        self.layBoutons = QHBoxLayout()
        self.layBoutons.addWidget(self.btnAll)
        #~ self.layBoutons.addWidget(self.btnNbAlbums)
        self.layBoutons.addWidget(self.btnAdd)
        self.layBoutons.addWidget(self.btnQuit)
        
        self.centralLayout.addLayout(self.layBoutons)
        
        #Connexion a la base et query initial
        connexion.openSqlConnection('QSQLITE', 'albums.db')
        self.query = "SELECT id, artist, album, year, genre, tracks, length FROM albums ORDER BY artist"
        
        #Setup du modele de la table (fixe)
        self.model = csm.CustomSqlModel()
        self.model.setQuery(self.query)
        column_names = self.model.getColumnNames()
        for i in range(len(column_names)):
            self.model.setHeaderData(i, Qt.Horizontal, QVariant(column_names[i]))
        
        #Setup de la table (fixe)
        self.tabView = QTableView()
        self.tabView.setModel(self.model)
        self.tabView.setColumnHidden(0, True) #Pas besoin d'afficher l'Id
        self.tabView.resizeColumnsToContents()
        self.centralLayout.addWidget(self.tabView)
        
        self.centralWidget.setLayout(self.centralLayout)

    def showAll(self):
        self.query = "SELECT id, artist, album, year, genre, tracks, length FROM albums ORDER BY artist"
        self.model.setQuery(self.query)

    def showNbAlbums(self):
        pass
        #~ self.query = "SELECT id, artiste, count(*), annee, genre, sum(pistes), time(sum(duree), 'unixepoch') FROM collection GROUP BY artiste ORDER BY artiste"
        #~ self.model.setQuery(self.query)
    
    def ajout(self):
        pass
        f = ajoutalbum.AjoutAlbum()
        f.exec_()
        self.model.refresh()
        #~ self.query = "SELECT id, artiste, album, annee, genre, pistes, time(duree, 'unixepoch') FROM collection ORDER BY artiste"
        #~ self.model.setQuery(self.query)

if __name__ == '__main__':
    app = QApplication(sys.argv)
    a = Vue(None)
    a.show()
    app.exec_()
